### PR TITLE
More updates for n:m support - battery pool's power formula

### PR DESCRIPTION
**Pull request by Marenz**
This specifically focuses on the power forumla for the battery pool.

* [x] test with no battery
* [x] test with no inverter
* [x] test with requests that don't specify all batteries behind the inverters

refs #501

### CODE REVIEW

File: src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py [hunk 1]
```diff
@@ -62,15 +65,38 @@ def generate(
...
 
         battery_inverters = list(
-            next(iter(component_graph.predecessors(bat_id))) for bat_id in component_ids
+            frozenset(
+                filter(
+                    component_graph.is_battery_inverter,
+                    component_graph.predecessors(bat_id),
+                )
+            )
+            for bat_id in component_ids
```
Comment (shsms) [inline src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py:74]: in a setup with 1 inverter and 2 batteries, there will be 2 frozensets with the same inverter.  that would lead to 2x power reporting, no?

File: src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py [hunk 2]
```diff
@@ -62,15 +65,38 @@ def generate(
...
+                "All batteries behind a set of inverters must be requested."
             )
 
-        for idx, comp in enumerate(battery_inverters):
+        builder.push_oper("(")
+        builder.push_oper("(")
+        # Iterate over the flattened list of inverters
+        for idx, comp in enumerate(
+            inverter for inverters in battery_inverters for inverter in inverters
+        ):
```
Comment (shsms) [inline src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py:99]: maybe the inverter ids can be collected into a set before enumerating, to deduplicate?

```suggestion
        for idx, comp in enumerate(
            set(inverter for inverters in battery_inverters for inverter in inverters)
        ):

File: src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py (hunk 1 above)
Comment (Marenz) [reply src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py:74]: Good catch. Added a test to catch this and fixed it, too

File: src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py (hunk 2 above)
Comment (Marenz) [reply src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py:99]: I made `battery_inverters` directly to a `frozenset` instead

File: src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py [hunk 3]
```diff
@@ -64,7 +64,7 @@ def generate(
 
         component_graph = connection_manager.get().component_graph
 
-        battery_inverters = list(
+        battery_inverters = frozenset(
```
Comment (shsms) [inline src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py:67]: was this supposed to go into a different commit?

Comment (Marenz) [reply src/frequenz/sdk/timeseries/formula_engine/_formula_generators/_battery_power_formula.py:67]: fixed >:|

**Review by shsms — APPROVED** [review]
